Requirements parsing code chokes on comments

it will probably also fail on empty lines and such

is also crashes on -e 

# example

```
self = <inmanta.env.VirtualEnv object at 0x7f8f63600dd8>
requirements_list = ['# ncdiff used from source due to a bug in latest release', '-e git://github.com/CiscoTestAutomation/ncdiff.git#egg=ncdiff', 'Jinja2==2.11.0', 'email_validator==1.0.5', 'lxml', 'ncclient', ...]

    def _remove_requirements_present_in_parent_env(self, requirements_list: List[str]) -> List[str]:
        reqs_to_remove = []
        packages_installed_in_parent = self.get_package_installed_in_parent_env()
        for r in requirements_list:
>           parsed_req = list(pkg_resources.parse_requirements(r))[0]
E           IndexError: list index out of range
````

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,7 @@
  - Fixed deadlock triggered by std::AgenConfigHandler (#1662)
  - Removed the resourceversionid table from the database (#1627)
  - Remote machines not being available or not having a python interpreter now results in a clearer error.
+ - Parse comments and urls correctly from the requirements.txt file of an Inmanta module (#1764)
 
 ## Added
 - Add get_environment_id to exporter (#1683)

--- a/src/inmanta/env.py
+++ b/src/inmanta/env.py
@@ -258,8 +258,12 @@ class VirtualEnv(object):
         """
             Install requirements from a list of requirement strings
         """
-        requirements_list = sorted(requirements_list)
+        # Install url-based requirements directly to prevent parsing issues
+        url_based_requirements = [r for r in requirements_list if "://" in r]
+        self._install(url_based_requirements)
 
+        # Install other requirements only when not already present in parent venv
+        requirements_list = sorted([r for r in requirements_list if "://" not in r])
         if detailed_cache:
             requirements_list = sorted(list(set(requirements_list) - self.__cache_done))
             if len(requirements_list) == 0:

--- a/src/inmanta/env.py
+++ b/src/inmanta/env.py
@@ -258,18 +258,12 @@ class VirtualEnv(object):
         """
             Install requirements from a list of requirement strings
         """
-        # Install url-based requirements directly to prevent parsing issues
-        url_based_requirements = [r for r in requirements_list if "://" in r]
-        self._install(url_based_requirements)
-
-        # Install other requirements only when not already present in parent venv
-        requirements_list = sorted([r for r in requirements_list if "://" not in r])
         if detailed_cache:
             requirements_list = sorted(list(set(requirements_list) - self.__cache_done))
             if len(requirements_list) == 0:
                 return
 
-        requirements_list = self._remove_requirements_present_in_parent_env(requirements_list)
+        requirements_list = sorted(self._remove_requirements_present_in_parent_env(requirements_list))
 
         # hash it
         sha1sum = hashlib.sha1()
@@ -290,6 +284,9 @@ class VirtualEnv(object):
         reqs_to_remove = []
         packages_installed_in_parent = self.get_package_installed_in_parent_env()
         for r in requirements_list:
+            # Always install url-based requirements
+            if "://" in r:
+                continue
             parsed_req = list(pkg_resources.parse_requirements(r))[0]
             # Package is installed and its version fits the constraint
             if (

--- a/src/inmanta/module.py
+++ b/src/inmanta/module.py
@@ -655,10 +655,43 @@ class Project(ModuleLike):
         """
             Collect the list of all python requirements off all modules in this project
         """
-        pyreq = [x.strip() for x in [mod.get_python_requirements() for mod in self.modules.values()] if x is not None]
-        pyreqa = "\n".join(pyreq).split("\n")
-        pyreqb = [x for x in pyreqa if len(x.strip()) > 0]
-        return list(set(pyreqb))
+        req_files = [x.strip() for x in [mod.get_python_requirements() for mod in self.modules.values()] if x is not None]
+        req_lines = [x for x in "\n".join(req_files).split("\n") if len(x.strip()) > 0]
+        req_lines = self._remove_comments(req_lines)
+        req_lines = self._remove_line_continuations(req_lines)
+        return list(set(req_lines))
+
+    def _remove_comments(self, lines: List[str]) -> List[str]:
+        """
+            Remove comments from lines in requirements.txt file.
+        """
+        result = []
+        for line in lines:
+            if line.strip().startswith("#"):
+                continue
+            if " #" in line:
+                line_without_comment = line.split(" #", maxsplit=1)[0]
+                result.append(line_without_comment)
+            else:
+                result.append(line)
+        return result
+
+    def _remove_line_continuations(self, lines: List[str]) -> List[str]:
+        """
+            Remove line continuation from lines in requirements.txt file.
+        """
+        result = []
+        line_continuation_buffer = ""
+        for line in lines:
+            if line.endswith("\\"):
+                line_continuation_buffer = f"{line_continuation_buffer}{line[0:-1]}"
+            else:
+                if line_continuation_buffer:
+                    result.append(f"{line_continuation_buffer}{line}")
+                    line_continuation_buffer = ""
+                else:
+                    result.append(line)
+        return result
 
     def get_name(self) -> str:
         return "project.yml"

--- a/tests/moduletool/test_python_dependencies.py
+++ b/tests/moduletool/test_python_dependencies.py
@@ -1,0 +1,56 @@
+"""
+    Copyright 2020 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import os
+
+import common
+from inmanta.module import Project
+
+
+def test_collect_python_requirements(tmpdir):
+    # Create project
+    common.makeproject(tmpdir, "test-project", deps=[("mod1", ""), ("mod2", "")], imports=["mod1", "mod2"])
+    project_dir = os.path.join(tmpdir, "test-project")
+    libs_dir = os.path.join(project_dir, "libs")
+    # Create mod1
+    common.makemodule(libs_dir, "mod1", project=False)
+    mod1 = os.path.join(libs_dir, "mod1")
+    mod1_req_txt = """iplib@git+https://github.com/bartv/python3-iplib
+pytest\
+>=\
+1.5
+
+
+iplib>=0.0.1
+    """
+    common.add_file(mod1, "requirements.txt", mod1_req_txt, msg="initial commit")
+    # Create mod2
+    common.makemodule(libs_dir, "mod2", project=False)
+    mod2 = os.path.join(libs_dir, "mod2")
+    mod2_req_txt = """# A comment
+dummy-yummy # A comment
+ # Another comment
+
+    """
+    common.add_file(mod2, "requirements.txt", mod2_req_txt, msg="initial commit")
+
+    project = Project(project_dir)
+    project.load_module("mod1")
+    project.load_module("mod2")
+    reqs = project.collect_python_requirements()
+    expected_reqs = ["iplib@git+https://github.com/bartv/python3-iplib", "pytest>=1.5", "iplib>=0.0.1", "dummy-yummy"]
+    assert sorted(reqs) == sorted(expected_reqs)

